Feat/corl 2474 allocate site mods

marcushaddon opened the pull request:
## What does this PR do?
This PR enables Organization Moderators to allocate Site Moderators to sites within their organization.

## What changes to the GraphQL/Database Schema does this PR introduce?
This updates the `UpdateUserRoleInput` type to accept an optional array of `siteIDs`.

## Does this PR introduce any new environment variables or feature flags? 
No.

## If any indexes were added, were they added to `INDEXES.md`?
n/a

## How do I test this PR?
1. As an organization moderator navigate to the Community view. You should see that UserRoleChange buttons are available for all users with roles <= moderator (so not for org mods or admins).
2. Click on the Role name of a non-moderator user. You should see that only the role Site Moderator is enabled.
3. Click on this role, and you should be able to add sites to this user in the following modal. Completing the modal should change that user's role to Site Moderator. 
4. Now that the user is a site mod, the only roles available to you in the RoleChange component should be Site Moderator and Commenter (this should also be true for users that were already Site Moderators when you arrived).
5. Set a site moderator back to a Commenter.
6. Log in as any other role and see that managing user roles from the Community view works (or doesn't work) just as before.
 
 
## How do we deploy this PR?
No special considerations should be required.

marcushaddon committed:
WIP: setup exhaustive testing of services/users.updateRole

marcushaddon committed:
WIP: setup exhaustive testing of services/users.updateRole

marcushaddon committed:
WIP: test services/users.updateRole

marcushaddon committed:
WIP: test services/users.updateRole

marcushaddon committed:
clean up test
delete outdated test

marcushaddon committed:
clean up test
delete outdated test

marcushaddon committed:
commit unsaved test

marcushaddon committed:
commit unsaved test

marcushaddon committed:
remove temp logging

marcushaddon committed:
remove temp logging

marcushaddon committed:
WIP: refactoring permissions

marcushaddon committed:
WIP: refactoring permissions

marcushaddon committed:
fix breaking changes after permissions refactor

marcushaddon committed:
fix breaking changes after permissions refactor

marcushaddon committed:
actually use refactored can() to determine if viewer can change users role

marcushaddon committed:
actually use refactored can() to determine if viewer can change users role

marcushaddon committed:
remove old permissions file

marcushaddon committed:
remove old permissions file

marcushaddon committed:
dont allow users < admin to allocate new admins

marcushaddon committed:
dont allow users < admin to allocate new admins

marcushaddon committed:
move validateRoleChange to common

marcushaddon committed:
move validateRoleChange to common

marcushaddon committed:
WIP: reafactoring common types

marcushaddon committed:
WIP: reafactoring common types

marcushaddon committed:
bring isSiteModerationScoped into common
fix bug in validateRoleChange + UserRoleChange

marcushaddon committed:
bring isSiteModerationScoped into common
fix bug in validateRoleChange + UserRoleChange

marcushaddon committed:
remove console.log

marcushaddon committed:
remove console.log

marcushaddon committed:
workaround for site mods in validateRoleChange

marcushaddon committed:
workaround for site mods in validateRoleChange

marcushaddon committed:
accept optional siteScopes with updateUserRole mutation

marcushaddon committed:
accept optional siteScopes with updateUserRole mutation

marcushaddon committed:
send siteIDs with updateRole mutation

marcushaddon committed:
send siteIDs with updateRole mutation

marcushaddon committed:
call correct mutation depending on user role

marcushaddon committed:
call correct mutation depending on user role

marcushaddon committed:
fix role name bug

marcushaddon committed:
fix role name bug

marcushaddon committed:
delete temp file

marcushaddon committed:
delete temp file

marcushaddon committed:
fix some linting errors

marcushaddon committed:
fix some linting errors

marcushaddon committed:
fix linting error

marcushaddon committed:
fix linting error

marcushaddon committed:
remove todos/bookmarks

marcushaddon committed:
remove todos/bookmarks

marcushaddon committed:
update failing snapshot

marcushaddon committed:
update failing snapshot

kabeaty commented on src/core/client/admin/components/UserRole/UserRoleChange.tsx:
```diff
@@ -6,6 +6,8 @@ import React, {
   useState,
 } from "react";
 
+import { validateRoleChange } from "coral-common/permissions";
```
Can you please remove the empty line below this to organize with the other coral imports?

marcushaddon replied:
this doesnt say outdated, but I'm not seeing an empty line below this import.

kabeaty commented on src/core/client/admin/components/UserRole/UserRoleChangeContainer.tsx:
```diff
@@ -68,8 +68,14 @@ const UserRoleChangeContainer: FunctionComponent<Props> = ({
   );
 
+  /**
+   * Q: MMODERATOR can changeRole requires there be further restrictions on
```
Did you get these questions answered and can remove this comment now?

kabeaty commented on src/core/client/admin/test/community/community-rtl.spec.tsx:
```diff
@@ -253,21 +253,6 @@ it("change user role", async () => {
   expect(resolvers.Mutation!.updateUserRole!.called).toBe(true);
 });
 
-it("can't change role as a moderator", async () => {
```
Instead of just removing this test, what do you think about updating it to test that roles can now be changed as expected when the viewer is an org moderator?

marcushaddon replied:
A new test has been added! src/core/client/admin/test/community/community-rtl.spec.tsx

kabeaty commented on src/core/common/permissions/types.ts:
```diff
@@ -0,0 +1,18 @@
+// TODO: this annoys me
```
Is this TODO resolved?

kabeaty commented on src/core/common/permissions/types.ts:
```diff
@@ -0,0 +1,18 @@
+export interface User {
+  id: string;
+  role: UserRole;
+  // tenantID: string;
```
Can this be removed?

kabeaty commented on src/core/common/permissions/validateRoleChange.ts:
```diff
@@ -0,0 +1,73 @@
+  user: Readonly<User>,
+  role: UserRole,
+  /**
+   * TODO: this optional argument that is only relelvant
```
Can you either remove this todo comment or update to remove typos for future reference on what we should do to change this?

kabeaty commented on src/core/server/models/user/helpers.ts:
```diff
@@ -4,14 +4,10 @@ import { SSOUserProfile } from "coral-server/app/middleware/passport/strategies/
 
 import { GQLUSER_ROLE } from "coral-server/graph/schema/__generated__/types";
 
+import { isSiteModerationScoped } from "coral-common/permissions";
```
Can you please include this import up with that from `coral-server` in the same grouping?

kabeaty commented on src/core/server/services/users/user.updateRole.spec.ts:
```diff
@@ -0,0 +1,209 @@
+import { Tenant } from "coral-server/models/tenant";
+import { User } from "coral-server/models/user";
+
+import { createMockMongoContex } from "coral-server/test/mocks";
```
Can you include this import with the above grouping from `coral-server`?

kabeaty commented on src/core/server/services/users/user.updateRole.spec.ts:
```diff
@@ -0,0 +1,209 @@
+    }
+  });
+
+  it("TODO: is this ok?! allows organization moderators to allocate site mods", async () => {
```
Did you want to update this todo? Looks like it's testing what we want it to, that org mods can add site mods?

kabeaty commented on src/core/server/services/users/user.updateRole.spec.ts:
```diff
@@ -0,0 +1,209 @@
+    }
+  });
+
+  it("does not allow site mods, staff, members, or commenters to change anyones role", async () => {
```
Site mods are able to make other users site mods for their sites and members though, right?

marcushaddon replied:
i don’t believe so, i think only organization moderators can do that. otherwise community mods for SBN could elevate their friends to moderators. i’ll confirm in standup tomorrow though

kabeaty replied:
I wanted to follow up on this. If you go into the `Community` tab in the admin, site moderators currently can change users' roles to site moderator or member for their site(s), so I think this test should be updated.

marcushaddon replied:
✅

kabeaty commented on src/core/server/services/users/users.ts:
```diff
@@ -757,19 +761,50 @@ export async function updateUsernameByID(
+  const validUpdate = validateRoleChange(fullViewer, user, role, isScoped);
+
+  if (!validUpdate) {
+    throw new UserForbiddenError("TODO: reason", "user", "updateRole");
```
Can you please update this reason?

kabeaty commented on src/core/client/admin/components/UserRole/UserRoleChange.tsx:
```diff
@@ -25,8 +28,8 @@ import UserRoleText from "./UserRoleText";
 interface Props {
-  username: string | null;
-  onChangeRole: (role: GQLUSER_ROLE_RL) => Promise<void>;
+  user: UserRoleChangeContainer_user;
```
Do we need to send through `user` separately here? It looks like we are still only using the info about the user that is already being passed through (`role` and `moderationScopes`) to validate the role changes, unless I'm missing something? 

If we do send `user` through like this, then I think we should update to not also send through `role`, `moderationScopes`, `membershipScopes`, `moderationScoped`, and `membershipScoped` separately.

kabeaty reviewed (dismissed):
Looks good! I left some comments throughout. I have a few overall questions on functionality, too. After I have an org mod promote someone to be a site mod, the option to then update them back to a `Commenter` is disabled. Shouldn't the org mod be able to make them a commenter again? 

Also, other role change dropdowns for admins and site mods don't include disabled roles like are displayed for org moderator (for example, Org Moderator, Admin, Staff, and Member are included in the org mod role change dropdown but grayed out as options). Is it intended that we want to include disabled options in this case? It also seems like an org moderator should be able to update a user to be a member?

I also see an empty spec file added (`UserRole.spec.tsx`). Can you please remove?

marcushaddon committed:
WIP responding to comments

marcushaddon committed:
WIP responding to comments

marcushaddon committed:
WIP updating tests

marcushaddon committed:
WIP updating tests

marcushaddon committed:
finish frontend test
fix backend test

marcushaddon committed:
finish frontend test
fix backend test

marcushaddon committed:
delete commented out lines

marcushaddon committed:
delete commented out lines

marcushaddon committed:
address comments

marcushaddon committed:
address comments

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon commented:
@kabeaty, I can hide rather than disable the invalid roles without much effort. I'd like to get input from Andrew on whether or not Org mods should be able to demote users back to commenters and / or allocate members. That empty test file should be deleted

marcushaddon committed:
create top level validatePermissionsAction
fix bug in validateScopeChange

marcushaddon committed:
create top level validatePermissionsAction
fix bug in validateScopeChange

marcushaddon committed:
WIP refactoring common permissions code

marcushaddon committed:
WIP refactoring common permissions code

marcushaddon committed:
fix updateUserRole tests

marcushaddon committed:
fix updateUserRole tests

marcushaddon committed:
fix tests

marcushaddon committed:
fix tests

marcushaddon committed:
remove eslint-disable

marcushaddon committed:
remove eslint-disable

marcushaddon commented on src/core/common/permissions/rules.ts:
```diff
@@ -0,0 +1,88 @@
+  reason: "Admins may change any non admin's role or scopes",
+});
+
+// org mods can promote anyone < site mod to be an unscoped member
```
This is a lot of comments, and was honestly to help me think, but I feel like this should be documented somewhere. Maybe I'll create a Slab doc and put a link to it here

marcushaddon commented on src/core/common/permissions/types.ts:
```diff
@@ -0,0 +1,120 @@
+import { isSiteModerationScoped } from "./isSiteModerationScoped";
```
A lot of the types here are redundant but are needed (as far as I know) because we can't shouldn't import + depend on types from the backend in common. Please correct me if wrong!

marcushaddon commented on src/core/common/permissions/validateRoleChange.ts:
```diff
@@ -0,0 +1,68 @@
+  user: Readonly<User>,
+  role: UserRole,
+  /**
+   * TODO: this optional argument that is only relevant
```
I intend to leave this TODO here until we break MODERATOR into SITE_MODERATOR + ORG_MODERATOR

marcushaddon commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -2779,6 +2779,12 @@ type UserModerationScopes {
+  sites (when not null) list the sites that the user is only allowed to moderate
+  on. When null, it means that the user is allowed to moderate on all sites.
+  """
+  siteIDs: [String!]
```
exposing these to the frontend because we have to compute these all over the place.

marcushaddon commented on src/core/server/services/users/permissions/index.ts:
```diff
@@ -0,0 +1,100 @@
+      );
+  }
+
+  return async () => null; // believe it or not it is yelling at me unless i do this
```
Currently looking into this. It is complaining that "not all code paths return a value" even though the return type is potentially undefined

marcushaddon committed:
clean up

marcushaddon committed:
clean up

marcushaddon committed:
WIP making org mod use correct component

marcushaddon committed:
WIP making org mod use correct component

marcushaddon committed:
WIP updating SiteRoleActionsModals

marcushaddon committed:
WIP updating SiteRoleActionsModals

marcushaddon committed:
WIP: refactoring SiteModeratorActionsModal, SiteRoleActionsSites

marcushaddon committed:
WIP: refactoring SiteModeratorActionsModal, SiteRoleActionsSites

marcushaddon committed:
resolve conflict

marcushaddon committed:
resolve conflict

marcushaddon committed:
go back to using UserRoleChange when viewer is org mod
hide inaccessible options instead of disabling
update tests

marcushaddon committed:
go back to using UserRoleChange when viewer is org mod
hide inaccessible options instead of disabling
update tests

marcushaddon committed:
WIP
hide inaccessible options
allow org mods to demote site mods to members
fixing tests

marcushaddon committed:
WIP
hide inaccessible options
allow org mods to demote site mods to members
fixing tests

marcushaddon committed:
revert SiteRoleActions to use original mutations

marcushaddon committed:
revert SiteRoleActions to use original mutations

marcushaddon committed:
open UpdateUserMembershipScopes to org mods

marcushaddon committed:
open UpdateUserMembershipScopes to org mods

marcushaddon committed:
make test name more accurate

marcushaddon committed:
make test name more accurate

marcushaddon committed:
Revert "make test name more accurate"

This reverts commit 2d63945814d0dd23f03bf3c2cf65da9dda306202.

marcushaddon committed:
Revert "make test name more accurate"

This reverts commit 2d63945814d0dd23f03bf3c2cf65da9dda306202.

marcushaddon committed:
remove comments

marcushaddon committed:
remove comments

marcushaddon committed:
update failing snapshot

marcushaddon committed:
Merge branch 'feat/CORL-2474-allocate-site-mods' of github.com:coralproject/talk into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
fix test

marcushaddon committed:
fix linting errors

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
fix small bug when changing role from member

marcushaddon committed:
WIP
use validatePermissionsAction consistently in SiteRoleActions

marcushaddon committed:
handle non scoped users in predicate

marcushaddon committed:
update tests
match order of options between site mod/org mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

nick-funk commented on src/core/client/admin/components/UserRole/SiteModeratorActionsModal.tsx:
```diff
@@ -70,68 +72,53 @@ const SiteModeratorActionsModal: FunctionComponent<Props> = ({
+                  <Localized id="community-siteModeratorsArePermitted">
+                    <ModalBodyText>
+                      Site moderators are permitted to make moderation
+                      suspensions on the sites they are assigned.
```
Has this been updated in `admin.ftl`?

marcushaddon replied:
oh i actually didnt mean to update this! changed it back

nick-funk commented on src/core/common/permissions/rules.ts:
```diff
@@ -0,0 +1,88 @@
+ * should be disallowed.
+ */
+export const validationRules: PermissionsActionRuleTest[] = [
+  adminsArePowerful,
```
this variable name 😂

nick-funk reviewed (changes requested):
Found one small localization that may be missing from an ftl file. Otherwise, I really like this PR and appreciate all the refactoring you've done to make things cleaner here. It's really great to see all the permissions logic having some sense applied to it. The `siteIDs` be available from the API to avoid re-computation everywhere is also a great idea.

marcushaddon committed:
replace accidentally deleted text

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

marcushaddon committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods

tessalt committed:
Merge branch 'develop' into feat/CORL-2474-allocate-site-mods
